tests/catalog/constraints_test.cpp

```diff
@@ -0,0 +1,264 @@
+    // Test1: insert 2 tuple with duplicated primary key
+    TransactionScheduler scheduler(2, data_table.get(), &txn_manager);
+    scheduler.Txn(0).Insert(10, 10);
+    scheduler.Txn(1).Insert(10, 11);
+    scheduler.Txn(0).Commit();
+    scheduler.Txn(1).Commit();
+
+    scheduler.Run();
+
+    EXPECT_TRUE((RESULT_SUCCESS == scheduler.schedules[0].txn_result
+        && RESULT_ABORTED == scheduler.schedules[1].txn_result) ||
+        (RESULT_SUCCESS == scheduler.schedules[1].txn_result
+            && RESULT_ABORTED == scheduler.schedules[0].txn_result));
+  }
+
+  {
+    std::unique_ptr<storage::DataTable> data_table(
+        TransactionTestsUtil::CreatePrimaryKeyUniqueKeyTable());
+    // Test2: update a tuple to be a illegal primary key
+    // txn1: update (1, 1) -> (1,11) -- success
+    // txn0: update (0, 0) -> (0,1) -- fail
+    // txn1 commit
+    // txn0 commit
+    TransactionScheduler scheduler(2, data_table.get(), &txn_manager);
+    scheduler.Txn(1).Update(1, 11);
+    scheduler.Txn(0).Update(0, 1);
+    scheduler.Txn(1).Commit();
+    scheduler.Txn(0).Commit();
+
+    scheduler.Run();
+
+    EXPECT_TRUE(RESULT_ABORTED == scheduler.schedules[0].txn_result);
+    EXPECT_TRUE(RESULT_SUCCESS == scheduler.schedules[1].txn_result);
+  }
+
+  {
+    std::unique_ptr<storage::DataTable> data_table(
+        TransactionTestsUtil::CreatePrimaryKeyUniqueKeyTable());
+    // Test3: update a tuple to be a legal primary key
+    // txn1: update (1, 1) -> (1,11) -- success
+    // txn1 commit
+    // txn0: update (0, 0) -> (0,1) -- success
+    // txn0 commit
+    TransactionScheduler scheduler(2, data_table.get(), &txn_manager);
+    scheduler.Txn(1).Update(1, 11);
+    scheduler.Txn(1).Commit();
+    scheduler.Txn(0).Update(0, 1);
+    scheduler.Txn(0).Commit();
+
+    scheduler.Run();
+
+    EXPECT_TRUE(RESULT_SUCCESS == scheduler.schedules[0].txn_result);
+    EXPECT_TRUE(RESULT_SUCCESS == scheduler.schedules[1].txn_result);
+  }
+}
+#endif
+
+#ifdef FOREIGHN_KEY_TEST
+TEST_F(ConstraintsTests, ForeignKeyInsertTest) {
+  // First, initial 2 tables like following
+  //     TABLE A -- src table          TABLE B -- sink table
+  // int(primary, ref B)  int            int(primary)  int
+  //    0                 0               0             0
+  //    1                 0               1             0
+  //    2                 0               2             0
+  //                                      .....
+  //                                      9             0
+
+  // create new db
+  auto &manager = catalog::Manager::GetInstance();
+  oid_t current_db_oid = bridge::Bridge::GetCurrentDatabaseOid();
+  auto newdb = new storage::Database(current_db_oid);
+  manager.AddDatabase(newdb);
+
+  auto table_B = TransactionTestsUtil::CreateTable(10, "tableB",0, 1001, 1001, true);
+  auto table_A = TransactionTestsUtil::CreateTable(3, "tableA", 0, 1000, 1000, true);
+
+  (void)table_A;
```

Comment:
I am trying to coax the compiler incase that table_A is not used later and gcc throw me warnning